### ISSUE TITLE
Question on TRN Randomness

### ISSUE THREAD

**Issue by xanderdunn** [aws-neuron/aws-neuron-sdk#676]
I have an `RngUniform` generating some random values between `[0, 1)` in an XLA graph run on TRN. It's generating 16 values:
```
Random values: [0.5, 0.5, 0.5, 0.5, 0.5078125, 0.515625, 0.53125, 0.5625, 0.515625, 0.53125, 0.5625, 0.625, 0.5234375, 0.546875, 0.59375, 0.6875]
```

This does not look random. For comparison, these are the values I see when running the same graph on CPU:
```
Random values: [0.59189475, 0.40608132, 0.8396015, 0.9122925, 0.0066410303, 0.86028945, 0.25914526, 0.20350039, 0.17746115, 0.8912425, 0.69978595, 0.2410804, 0.23565137, 0.9759909, 0.6240233, 0.627475]
```

Here are the NEFF and XLA HLO .pb of this graph: 
[randomness_4206306573283728891.neff.zip](https://github.com/aws-neuron/aws-neuron-sdk/files/11457175/randomness_4206306573283728891.neff.zip), 
[rust_hlo_randomness.pb.zip](https://github.com/aws-neuron/aws-neuron-sdk/files/11457176/rust_hlo_randomness.pb.zip)

Is it expected that the first handful of values should be identical and then all close to 0.5? I find that when I generate a large number of random values, for example 1M, they do appear to be random in large quantity. 

In addition, is there a way for me to set a seed in the XLA graph such that random values are different between runs? I saw in the [compiler release notes](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/release-notes/compiler/neuronx-cc/index.html?highlight=seed) that the `RngBitGenerator` ignores user seed, but this could be outdated, this apparently doesn't apply to `RngUniform`, and I haven't figured out how to set the user seed.

**Comment by aws-diamant** [comment]
hey @xanderdunn !
we're aware of this issue and actively debugging it, it'll likely require a firmware update to fix (there will be no perf impact).
there is a way to seed the XLA graphs, yes.

let me circle back with the team on a few details, and update you asap?
/Ron

**Comment by xanderdunn** [comment]
That sounds great, thank you!

**Comment by xanderdunn** [comment]
1 - Great to hear!
2 - Thanks, we actually don't use PyTorch or even any Python at all. We have a custom Rust framework for constructing our models, so we output the XLA HLO ourselves. I'm seeing [set_rng_state](https://github.com/pytorch/xla/blob/30b2aba0ec2d9f7c664025f805bf4677c0f8a13e/torch_xla/core/xla_model.py#L1056) in PyTorch XLA, but unsure how this is translated to XLA HLO. I don't see a particular location in the [xla_data.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/xla_data.proto) or [hlo.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/hlo.proto) for setting rng seed or state. One way to help us here would be to provide a simple XLA HLO .pb file that sets the rng state. We could inspect that .pb to figure out how to construct it from our Rust framework.
3 - We do our recomputation of the forward on different machines. I suppose we could communicate the cached dropout masks from machine to machine, but there'd be less cache and communication overhead if we can get 2 right.

**Comment by xanderdunn** [comment]
I upgraded to Neuron SDK 2.11 and the randomness appears to be fixed.

Generating 16 values:
```
Random values: [0.76087344, 0.521747, 0.36900663, 0.08698821, 0.37609327, 0.75218666, 0.82928145, 0.008746624, 0.5306226, 0.061245203, 0.28928924, 0.24498105, 0.14590251, 0.29180503, 0.7507657, 0.16722012]
```

The Kolmogorov-Smirnov test is passing on all of our sample sizes now.

Thanks very much for the bug fix!

**Comment by xanderdunn** [comment]
One thing we noticed in 2.11 is that the performance of RngUniform seems to be much worse, perhaps as a result of this change or some other change in 2.11.

We had a Transformer benchmark with inputs generated in-graph via RngUniform that went from 0.13111s average over 20 runs to 0.8866s. Generating the random inputs outside of the Graph and feeding the values to the model brought this back to an expected performance.
